### PR TITLE
Add s3 support, fix bug with chrome.go, fix readme curl example.

### PR DESCRIPTION
**Pull request by mdanzinger**
Hey! Sorry for the delay in getting the PR your way. I've added an `s3Object` which implements your `StoredFile` interface. I've used a singleton pattern to ensure we're not creating new s3 clients on every new `s3Object` creation.

Code can probably be cleaned up / simplified - looking forward to whatever feedback you may have!

Thanks!

### CODE REVIEW

File: renderer/chrome.go [hunk 1]
```diff
@@ -94,6 +92,7 @@ func CreatePdf(ctx context.Context, params ChromeParameters) ([]byte, []byte, er
 	newContextTarget, err := baseBrowser.Target.CreateBrowserContext(ctx)
 	if err != nil {
 		return nil, nil, err
+		log.Printf("BBO")
```
Comment (vjiandani-r7) [inline renderer/chrome.go]: Clean up

File: renderer/chrome.go [hunk 2]
```diff
@@ -63,18 +63,16 @@ func listenForResponse(c chan *network.ResponseReceivedReply, responseReceivedCl
...
-		pt, err = devt.Create(ctx)
-		if err != nil {
-			return nil, nil, err
-		}
+		return nil, nil, err
 	}
-	defer devt.Close(ctx, pt)
+
+
+
```
Comment (vjiandani-r7) [inline renderer/chrome.go]: Clean up

File: storage/s3.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+)
+
+type s3Client struct {
+	client *s3.S3
+	uploader *s3manager.Uploader
+	downloader *s3manager.Downloader
+	bucket string
+}
+
+
```
Comment (vjiandani-r7) [inline storage/s3.go]: Clean up

File: README.md [hunk 4]
```diff
@@ -77,7 +82,7 @@ curl -X POST \
   -H 'Content-Type: application/json' \
   -d '{
 	"correlationId": "7667e2de-2f21-4ab7-9afb-402de2a6468f",
-	"fileName": "7667e2de-2f21-4ab7-9afb-402de2a6468f.zip"
+	"fileName": "7667e2de-2f21-4ab7-9afb-402de2a6468f.zip",
```
Comment (vjiandani-r7) [inline README.md:85]: Thanks

File: renderer/chrome.go [hunk 5]
```diff
@@ -63,18 +63,16 @@ func listenForResponse(c chan *network.ResponseReceivedReply, responseReceivedCl
 
 func CreatePdf(ctx context.Context, params ChromeParameters) ([]byte, []byte, error) {
 	// Use the DevTools API to manage targets
-	devt := devtool.New("http://127.0.0.1:9222")
-	pt, err := devt.Get(ctx, devtool.Page)
+	devt, err := devtool.New("http://127.0.0.1:9222").Version(ctx)
```
Comment (vjiandani-r7) [inline renderer/chrome.go:67]: I'm not familiar with what Version is doing. What's the reasoning behind this change?

File: storage/s3.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+// Variables needed to enforce singleton pattern for S3Client
+var (
+	client *s3Client
+	err error
```
Comment (vjiandani-r7) [inline storage/s3.go]: Maybe call this s3Err for disambiguation?
Edit: or whatever you want to call it.

File: renderer/chrome.go (hunk 5 above)
Comment (mdanzinger) [reply renderer/chrome.go:67]: I was getting an error `cdp.Target: CreateBrowserContext: rpc error: Not allowed. (code = -32000)` when attempting to render a page. I found this issue (https://github.com/mafredri/cdp/issues/80) where the author suggested using `.Version` which seemed to have fixed my issue.

File: storage/s3.go (hunk 3 above)
Comment (mdanzinger) [reply storage/s3.go]: Just confirming by "Clean up" you're referring to white space / formatting?

File: renderer/chrome.go (hunk 2 above)
Comment (mdanzinger) [reply renderer/chrome.go]: Just confirming by "Clean up" you're referring to white space / formatting?

File: storage/s3__test.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+func TestNewS3Client(t *testing.T) {
+	_, err := newS3Client()
+	if err != nil {
+		fmt.Println(err)
+		t.Fail()
+	}
+}
+
+func TestNewS3Object(t *testing.T) {
```
Comment (vjiandani-r7) [inline storage/s3__test.go:18]: Might as well test `Exists`?

File: storage/storage.go [hunk 8]
```diff
@@ -22,17 +22,21 @@ type StoredFile interface {
...
-	} else if storageStrategy == "disk" {
+	case "disk":
 		return &encryptedFile{
 			filePath: cfg.Config().StorageDirectory(),
 			fileName: fileName,
 		}, nil
-	} else {
+
+	case "s3":
+		return NewS3Object(fileName)
```
Comment (vjiandani-r7) [inline storage/storage.go:33]: Your way is fine but I would prefer consistency. Either add `New*` to the other two or do this the way they're done. Whichever you prefer.

**Comment by vjiandani-r7** [comment]
Looks good! Thanks for this. I'll wait until you give me the thumbs up to do a quick test locally. Assuming all is well, I'll approve/merge.

File: storage/s3.go (hunk 3 above)
Comment (vjiandani-r7) [reply storage/s3.go]: yes

File: renderer/chrome.go (hunk 2 above)
Comment (vjiandani-r7) [reply renderer/chrome.go]: yes

File: storage/s3.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+		s3client:client,
+	}, nil
+
+}
+
+func (o *s3Object) FileName() string {
+	return o.fileName
+}
+
+func (o *s3Object) Write(data []byte) error {
```
Comment (vjiandani-r7) [inline storage/s3.go:102]: I'll document somewhere that the behavior for Write is to overwrite whatever is there... This is true for all the storage strategies.

File: storage/storage.go (hunk 8 above)
Comment (mdanzinger) [reply storage/storage.go:33]: Do you have any preference? I opted to use the factory pattern because I needed to attach an s3 client to the object.

**Comment by mdanzinger** [comment]
Thanks for taking the time to review this. I'll be implementing your changes this weekend.

File: storage/storage.go (hunk 8 above)
Comment (vjiandani-r7) [reply storage/storage.go:33]: No preference.

File: storage/s3.go (hunk 9 above)
Comment (mdanzinger) [reply storage/s3.go:102]: Currently `o.Write` will override the s3 object - I've added a test case for that.

File: storage/storage.go (hunk 8 above)
Comment (mdanzinger) [reply storage/storage.go:33]: I've gone ahead and made constructors for the other storage options too, it just seemed a bit too coupled to be attaching an s3 client there.

File: storage/s3__test.go (hunk 7 above)
Comment (mdanzinger) [reply storage/s3__test.go:18]: Added in latest push

**Comment by mdanzinger** [comment]
@vjiandani-r7 Pushed changes as per your feedback.

File: storage/s3.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	bucketInput := &s3.CreateBucketInput{
+		Bucket: aws.String(cfg.Config().S3Bucket()),
+	}
+
+	_, err := svc.CreateBucket(bucketInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyExists:
+				fmt.Printf("%s bucket already exits", cfg.Config().S3Bucket())
```
Comment (vjiandani-r7) [inline storage/s3.go]: Just noticed. Please use the logger for all these print outs.

Example:
```log "github.com/sirupsen/logrus"```
```
		log.Debug(
			fmt.Sprintf(
				"Alloc = %v MiB, TotalAlloc = %v MiB, Sys = %v MiB, NumGC = %v",
				bToMb(m.Alloc),
				bToMb(m.TotalAlloc),
				bToMb(m.Sys),
				m.NumGC,
			),
		)
```

File: storage/s3.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+		Bucket: aws.String(o.s3client.bucket),
+		Key:    aws.String(o.FileName()),
+	}
+
+	_, err := o.s3client.client.GetObject(input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchKey:
+				fmt.Println(s3.ErrCodeNoSuchKey, aerr.Error())
```
Comment (vjiandani-r7) [inline storage/s3.go]: Logger

File: cfg/environment.go [hunk 12]
```diff
@@ -121,3 +122,11 @@ func (c *envConfig) PrintDeadline() time.Duration {
 
 	return printDeadline
 }
+
+func (c *envConfig) S3Bucket() string {
+	if v := os.Getenv("PDF_RENDERER_S3_BUCKET"); len(v) != 0 {
+		return  v
+	}
+	log.Fatal("Environment variable PDF_RENDERER_S3_BUCKET must be set.")
```
Comment (vjiandani-r7) [inline cfg/environment.go]: If this is fatal, we should check that it's a valid value before starting the web server and exit if it's  invalid.

**Comment by vjiandani-r7** [comment]
@mdanzinger I noticed a couple more things.

File: cfg/environment.go (hunk 12 above)
Comment (mdanzinger) [reply cfg/environment.go]: "Valid value" being a valid s3 name? `log.Fatal` should call `os.Exit(1)` under the hood, did you want this check done before the server starts? If so, any objection to tossing this check inside of the cfg package init?

Comment (vjiandani-r7) [reply cfg/environment.go]: No objection.

Comment (vjiandani-r7) [reply cfg/environment.go]: It should just have a non-empty, non-whitespace value.

**Comment by mdanzinger** [comment]
Hey @vjiandani-r7, I've gone ahead and implemented your changes.

File: cfg/environment.go (hunk 12 above)
Comment (mdanzinger) [reply cfg/environment.go]: I've put the `log.Fatal` inside of the init.

File: storage/s3.go (hunk 10 above)
Comment (mdanzinger) [reply storage/s3.go]: Out of curiosity,  does `fmt.Sprintf` inside of the `log.Debug` have any benefit over simply using `log.Debugf`?

Comment (vjiandani-r7) [reply storage/s3.go]: Not as far as I know. Feel free to use `log.Debugf`. I didn't have any specific reason for doing it this way.

Comment (mdanzinger) [reply storage/s3.go]: No worries, I'll keep the `fmt.Sprintf` for consistency.

**Comment by vjiandani-r7** [comment]
I'm getting an error when trying to render a pdf: ```Failed to render: https://google.com\n Error: cdp.Target: CreateBrowserContext: rpc error: Not allowed. (code = -32000)``` Looking but if you know, let me know.

**Comment by vjiandani-r7** [comment]
Might be related to that change you made for the chrome version issue.

**Comment by mdanzinger** [comment]
@vjiandani-r7 That was the error I was receiving too, commit https://github.com/rapid7/pdf-renderer/pull/6/commits/67acb5664a9e1a327904b1bf5bc18c02a7444b38 seemed to have fixed it for me. Are you running an updated version of chrome?

**Comment by vjiandani-r7** [comment]
I figured that one out. My bad on how I had things set up. I am running into the following issue while running the unit tests though:
```
go test -v ./...
?   	github.com/rapid7/pdf-renderer	[no test files]
?   	github.com/rapid7/pdf-renderer/cfg	[no test files]
?   	github.com/rapid7/pdf-renderer/correlation	[no test files]
?   	github.com/rapid7/pdf-renderer/renderer	[no test files]
=== RUN   TestNewS3Client
time="2019-02-19T12:50:55-08:00" level=error msg="MissingRegion: could not find region configuration"
MissingRegion: could not find region configuration
--- FAIL: TestNewS3Client (0.00s)
=== RUN   TestNewS3Object
time="2019-02-19T12:50:55-08:00" level=error msg="MissingRegion: could not find region configuration"
MissingRegion: could not find region configuration
--- FAIL: TestNewS3Object (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x13caa57]

goroutine 6 [running]:
testing.tRunner.func1(0xc4201cc1e0)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:742 +0x29d
panic(0x1438fa0, 0x1777c20)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/panic.go:502 +0x229
github.com/rapid7/pdf-renderer/storage.(*s3Object).Write(0x0, 0xc42001eea0, 0x11, 0x20, 0x11, 0x20)
	/r7/src/go-projects/src/github.com/rapid7/pdf-renderer/storage/s3.go:105 +0x47
github.com/rapid7/pdf-renderer/storage.TestNewS3Object(0xc4201cc1e0)
	/r7/src/go-projects/src/github.com/rapid7/pdf-renderer/storage/s3__test.go:25 +0xfa
testing.tRunner(0xc4201cc1e0, 0x14e71d0)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:777 +0xd0
created by testing.(*T).Run
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:824 +0x2e0
FAIL	github.com/rapid7/pdf-renderer/storage	0.075s
?   	github.com/rapid7/pdf-renderer/web	[no test files]
make: *** [test] Error 1
```

**Comment by mdanzinger** [comment]
Do you have a region set in the `~/.aws/credentials` file?

``` region = us-east-2 ```

**Comment by vjiandani-r7** [comment]
Ok I think this is fine. I'll probably want to add support for https://docs.aws.amazon.com/sdk-for-go/api/aws/ec2metadata/ though. Unless you would like to do that.

**Pull request merged**